Add support for Key-Value pairs

ellttBen opened the pull request:
Hi there,
As discussed in #329, this PR implements support for the `log::kv` structured logging API.
These are some user-facing choices I have made:
- Everything is implemented behind a `log_kv` feature
- Support is added in the two encoders `json` and `pattern`:
    - The `json` message has an `attributes` field which is a serialized map of the record's structured logs
    - The `pattern` encoder has a new `K` or `key_value` formatter which functions exactly like the `mdc` formatter.

If this sounds okay, I can also edit the documentation.

ellttBen commented on Cargo.toml:
```diff
@@ -26,13 +35,21 @@ delete_roller = []
 simple_writer = []
 threshold_filter = []
 background_rotation = []
+log_kv = ["log/kv"]
```
This is the only change in this file, the rest is formatting.

bconn98 commented on src/encode/json.rs:
```diff
@@ -163,6 +168,49 @@ impl Deserialize for JsonEncoderDeserializer {
+mod kv {
+    use log::kv::VisitSource;
+
+    pub fn get_attributes(source: &dyn log::kv::Source) -> anyhow::Result<Vec<(String, String)>> {
```
Please document pub fns

ellttBen replied:
I have marked this function as `pub(crate)` since it's not relevant to users.

bconn98 commented on src/encode/json.rs:
```diff
@@ -163,6 +168,49 @@ impl Deserialize for JsonEncoderDeserializer {
+        Ok(LogKvVisitor::process(source)?.yield_values())
+    }
+    #[derive(Default)]
+    struct LogKvVisitor {
```
Why a custom struct vs BTreeMap like the example in log::kv?

ellttBen replied:
I am a bit on the fence about this. My initial reasoning is that a Vec is more memory efficient, which enables us to minimize overall footprint of the system. However, looking at the `kv` documentation there is an explicit mention of the fact that
`A source doesn’t have to guarantee any ordering or uniqueness of key-values. If the given visitor returns an error then the source may early-return with it, even if there are more key-values.`.
This means that we need to guarantee uniqueness in order to produce valid JSON (`serde_json` also doesn't enforce uniqueness AFAIK).
So I think you're right and I'll switch it to a BTreeMap.

bconn98 commented on src/encode/json.rs:
```diff
@@ -163,6 +168,49 @@ impl Deserialize for JsonEncoderDeserializer {
+
 #[cfg(test)]
 #[cfg(feature = "simple_writer")]
 mod test {
```
Please add tests for your new type (if you keep it) and the ser_map method

ellttBen replied:
The new type was simplified to a BTreeMap which can only be used within the `get_attributes` function.

bconn98 commented on src/encode/pattern/mod.rs:
```diff
@@ -536,6 +536,47 @@ impl<'a> From<Piece<'a>> for Chunk {
+                        Some(arg) => {
+                            if let Some(arg) = arg.first() {
+                                match arg {
+                                    Piece::Text(key) => key.to_owned(),
```
Value not key

ellttBen replied:
This is basically the MDC code from just above, do you want me to also fix it there?

ellttBen replied:
In the meantime I have edited this reference but not the other one.

bconn98 commented on src/encode/pattern/mod.rs:
```diff
@@ -1031,6 +1083,46 @@ mod tests {
 
+    #[test]
+    #[cfg(all(feature = "simple_writer", feature = "log_kv"))]
+    fn kv() {
```
tests look good, can you also check some of the error cases? Take a look at #359 if you need some help

ellttBen replied:
I adapted the `test_from_piece_to_chunk` test from your PR, thanks!

bconn98 reviewed (commented):
Looks great! Lets try and get some more tests in there and the documentation

bconn98 commented on src/encode/json.rs:
```diff
@@ -162,6 +171,34 @@ impl Deserialize for JsonEncoderDeserializer {
+    pub(crate) type Map = BTreeMap<String, String>;
+
+    pub(crate) fn get_attributes(source: &dyn log::kv::Source) -> anyhow::Result<Map> {
+        struct Visitor {
+            inner: Map,
+        }
+        impl<'kvs> VisitSource<'kvs> for Visitor {
+            fn visit_pair(
+                &mut self,
+                key: log::kv::Key<'kvs>,
+                value: log::kv::Value<'kvs>,
+            ) -> Result<(), log::kv::Error> {
+                self.inner.insert(format!("{key}"), format!("{value}"));
+                Ok(())
+            }
+        }
```
Why have the struct and impl defined within a function?

ellttBen replied:
My reasoning was that they are implementation details and do not serve a purpose outside of this function.

estk replied:
@ellttBen Please move the impl out of the fn definition into the `kv` namespace.

bconn98 commented on src/encode/pattern/mod.rs:
```diff
@@ -751,6 +812,7 @@ mod tests {
     #[cfg(feature = "simple_writer")]
     use std::thread;
 
+    use super::*;
     use super::{Chunk, PatternEncoder};
```
Wouldn't keep both of these

ellttBen commented:
Hi, sorry for the delay, I'm still keen to get this merged. I have rebased this PR and squashed the commits. If more documentation is needed I can get that in ASAP as well.

bconn98 commented:
I'll take a look at this soon. I'll do my best to get a hold of the project owner and work through our recent batches of MRs

bconn98 commented on Cargo.toml:
```diff
@@ -27,13 +36,21 @@ fixed_window_roller = []
 simple_writer = []
 threshold_filter = []
 background_rotation = []
+log_kv = ["log/kv"]
```
Looks like log::kv is unstable in the current requested log version. Mind fixing this?

sdroege commented:
What's the status of this PR?

bconn98 commented:
@sdroege I am targeting next week to start working with the owner of the repo. Appreciate everyone's hard work on the recent batch of MRs and can't wait to get things moving again!

sdroege commented:
That sounds great, thanks!

estk commented on src/encode/json.rs:
```diff
@@ -162,6 +171,34 @@ impl Deserialize for JsonEncoderDeserializer {
+    pub(crate) fn get_attributes(source: &dyn log::kv::Source) -> anyhow::Result<Map> {
+        struct Visitor {
+            inner: Map,
+        }
```
Please move this struct defn out of the fn. Additionally, for performance lets add a `with_capacity` method that allows the user to do:
```
let visitor = Visitor::with_capacity(source.count());
```

Optionally visitor might be better as a tuple variant: `struct Visitor(Map)`

ellttBen replied:
The underlying [BTreeMap](https://doc.rust-lang.org/stable/std/collections/struct.BTreeMap.html) does not have a `with_capacity` method, so I'm not sure what a `with_capacity` would look like for Visitor.

estk replied:
Oh, yep you're right. Force of habit to always want to with_capacity stuff, apologies for the error. While we're on the subject, what is the reasoning to use a BTree instead of a HashMap or Vec<(String, String)> that allows for preallocation?

ellttBen replied:
I initially just thought it would be a more minimal memory footprint, with little overhead as my understanding is that log attributes are often not too numerous. However this is all basically moot now, I've switched to direct serialization from the log::kv::Source object with dynamic dispatch, following your recommendation.

estk commented on src/encode/pattern/mod.rs:
```diff
@@ -73,6 +73,14 @@
+//!     to the empty string. This formatter requires the `log_kv` feature to be
+//!     enabled.
+//!     * `{K(user_id)}` - `123e4567-e89b-12d3-a456-426655440000`
+//!     * `{K(nonexistent_key)(no mapping)}` - `no mapping`
```
nice doc!

ellttBen replied:
Thanks!

estk commented on src/encode/pattern/mod.rs:
```diff
@@ -536,6 +545,47 @@ impl<'a> From<Piece<'a>> for Chunk {
                     }
                 }
+                #[cfg(feature = "log_kv")]
+                "K" | "key_value" => {
```
Optional: We may want to warn the user on parse that the `log_kv` feature is not enabled if the user attempts to use `K`

estk commented on src/encode/pattern/mod.rs:
```diff
@@ -536,6 +545,47 @@ impl<'a> From<Piece<'a>> for Chunk {
+                            }
+                        }
+                        None => "",
+                    };
```
Id like to see this and MDC stuff get factored into a helper function rather than a copy-paste.

estk commented on src/encode/pattern/mod.rs:
```diff
@@ -751,7 +812,7 @@ mod tests {
     use std::thread;
 
-    use super::{Chunk, PatternEncoder};
+    use super::*;
```
I prefer not to use * imports please import items directly.

estk reviewed (commented):
I left a couple of comments, but looking really good.

One more general question, did you attempt to wrap `log::kv::Source` in a newtype and impl serde::Serialize on it?

Just wondering because I was thinking about the tradeoffs of allocating a whole BTree vs the annoyance of adding lifetimes to `Message`.

ellttBen commented:
I implemented the suggested review changes, and also changed the approach for json encoding to directly (via a new wrapper type) serializing from the &log::kv::Source dynamic reference. The error handling is a bit finicky with transformations between log::kv::Error and the serializer errors, but I think overall it makes more sense.

estk reviewed (approved):
nailed it

estk commented:
blocked on #367

SL-RU commented:
Hi! What is the current status of the MR?

bconn98 commented:
> Hi! What is the current status of the MR?

@SL-RU we're still waiting on the blocked PR

SL-RU commented on Cargo.toml:
```diff
@@ -1,7 +1,10 @@
 name = "log4rs"
 version = "1.3.0"
-authors = ["Steven Fackler <sfackler@gmail.com>", "Evan Simmons <esims89@gmail.com>"]
+authors = [
```
I think you need to revert changes in the formatting

ellttBen replied:
Changes have been approved by the maintainers, I'm reluctant to push to the PR at this stage for that reason

SL-RU commented:
Hi! What is the current status of the MR?

bconn98 commented:
Hey! Sorry but all PRs are blocked by the PR fixing the deprecated API. I've been too busy of late to bug @estk and get it moving again. Please feel free to take a look at that PR and suggest options you might see in helping bring it to closure.

SL-RU commented:
Hi, what is current status? What blocks this PR now?

SL-RU commented:
It seems repo was abandoned. Isn't it? 
Almost half of year without progress?

SL-RU commented:
> @SL-RU yes, as far as I'm concerned this project is abandoned. The owner, was not happy with the solution to resolve #367 which blocks all further PRs. Until such a time as he turns the project over to be community driven, I don't foresee a solution to grow the project.

Thank you for the reply, do you know some project with same functionality?

bconn98 commented:
https://github.com/estk/log4rs/issues/412

estk commented:
Bryan had agreed to send me an email when he saw PRs as ready. My last email from him was Feb last year.

i am adding this review to my queue.

estk commented:
@ellttBen can you please rebase?
